### ISSUE TITLE
Admin user asked interactively sometimes, using mysql and persistence

### ISSUE THREAD

**Issue by pedroetb** [ChatDeBlofeld/seafile-arm-docker-base#20]
Sometimes, I'm getting following log output:

```
2021-10-31T14:49:31+00:00 [Entrypoint] Local time set to Atlantic/Canary
2021-10-31T14:49:31+00:00 [Entrypoint] Checking permissions
2021-10-31T14:49:31+00:00 [Entrypoint] PUID not set, using current
2021-10-31T14:49:32+00:00 [Entrypoint] GUID not set, using current
2021-10-31T14:49:32+00:00 [Entrypoint] Adjusting identifiers
usermod: no changes
2021-10-31T14:49:35+00:00 [Entrypoint] Running launch script
2021-10-31T14:49:37+00:00 [Launch] Making symlink to latest version
2021-10-31T14:49:37+00:00 [Launch] Linking internal configuration and data folders with the volume
2021-10-31T14:49:37+00:00 [Launch] Launching seafile
Starting seafile server, please wait ...
** Message: 14:49:40.529: seafile-controller.c(536): No seafevents.
Seafile server started
Done.
LANG is not set in ENV, set to en_US.UTF-8
./seafile-server-latest/seahub.sh: line 210: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
LC_ALL is not set in ENV, set to en_US.UTF-8
Starting seahub at port 8000 ...
----------------------------------------
It's the first time you start the seafile server. Now let's create the admin account
----------------------------------------
What is the email for the admin account?
[ admin email ] 
Error happened during creating seafile admin.
Seahub is started
Done.
2021-10-31T14:50:23+00:00 [Launch] Done
2021-10-31T14:50:23+00:00 [Entrypoint] Waiting for termination
```

These lines in particular:

```
...
Starting seahub at port 8000 ...
----------------------------------------
It's the first time you start the seafile server. Now let's create the admin account
----------------------------------------
What is the email for the admin account?
[ admin email ] 
Error happened during creating seafile admin.
Seahub is started
...
```

Also, sometimes only shows these (without interactive messages, but with the error line):

```
...
Starting seahub at port 8000 ...
Error happened during creating seafile admin.
Seahub is started
...
```

I'm not sure about the cause, but it appears randomly. Service is properly configured to use mysql/mariadb database, with admin user and pass set at environment variables and using volumes for persistence. My deployment configuration is published at: <https://gitlab.com/pedroetb-projects/seafile/-/blob/master/docker-compose.yml> (`docker-compose.yml` and `.env` for defaults).

Despite this, service is working fine, thanks! Only sharing this because it may be helpful.

**Comment by ChatDeBlofeld** [comment]
Hi, I've had this problem too sometimes but I never thought it could be blocking because of the interactive input. I guess you deleted the admin credentials from your dotenv (a good thing I don't do...) which triggers the interactive prompt.

Thanks for the report by the way, I think it's time to do something about it. 

FYI, this is probably because the database is not ready to accept connections when Seafile starts. This could be tedious to fix properly, so I left it out...

**Comment by pedroetb** [comment]
Actually, I'm not deleting admin credentials. Seafile and MariaDB are being deployed as Docker Swarm services, with these variables set as environment variables, so they are always available.

While testing, database was always available to Seafile, because I was restarting only Seafile service. It's strange that this happens randomly (1 of 5 retries, approximately).

I will continue testing with next releases, thanks for fast answering!

**Comment by ChatDeBlofeld** [comment]
> Actually, I'm not deleting admin credentials. Seafile and MariaDB are being deployed as Docker Swarm services, with these variables set as environment variables, so they are always available.

Yes you're right, this has nothing to do. I spoke too quickly.

> While testing, database was always available to Seafile, because I was restarting only Seafile service. It's strange that this happens randomly (1 of 5 retries, approximately).

Ok, I can reproduce (without blocking though) when database isn't reachable but this, it's totally something else.
Thus I have no idea what's happening. Could it be related to swarm usage ? I have very little experience with it.

**Comment by pedroetb** [comment]
I'm pretty sure Swarm is working fine here, database is reachable (through internal Docker network, on same machine or another one, the error may appear) while Seafile container is starting. I'm thinking about 2 possibilities:

1. Maybe is a Seafile bug, with a race condition when credentials are checked.
2. Or may be related with usage of `docker_entrypoint.sh`, `init.sh` or `launch.sh` at this repository, and some action made inside `init.sh` should be done always and not only on first boot.

This is just a guess, I haven't checked enough to be sure.

**Comment by ChatDeBlofeld** [comment]
Thanks for your time and your hypotheses.

At this point I have no idea on how reproduce this. The above commit mitigate the only behavior I can test (i.e database not reachable) which will probably not help but, well, image will be on the hub soon so you'll see. ^^

**Comment by pedroetb** [comment]
Thank you!

Ok, I will check log output when using next releases, no problem.

**Comment by olof-nord** [comment]
I am unfortunately also facing this problem now, with the following log output, from what I can tell the same problem as @pedroetb mentions.

Should the "seafile" user be created by Seafile, or exist before the init script?
I understand that the Seafile databases (ccnet, seafile and seahub) are per default created during setup.

```
seafile       | 2022-01-23T17:46:59+01:00 [Entrypoint] Local time set to Europe/Berlin
seafile       | 2022-01-23T17:46:59+01:00 [Entrypoint] Checking permissions
seafile       | 2022-01-23T17:46:59+01:00 [Entrypoint] PUID not set, using current
seafile       | 2022-01-23T17:46:59+01:00 [Entrypoint] GUID not set, using current
seafile       | 2022-01-23T17:46:59+01:00 [Entrypoint] Adjusting identifiers
seafile       | usermod: no changes
seafile       | 2022-01-23T17:46:59+01:00 [Entrypoint] Running launch script
seafile       | 2022-01-23T17:46:59+01:00 [Launch] Making symlink to latest version
seafile       | 2022-01-23T17:46:59+01:00 [Launch] Linking internal configuration and data folders with the volume
seafile       | 2022-01-23T17:46:59+01:00 [Launch] Waiting for db
seafile       | 2022-01-23T17:47:00+01:00 [Launch] Launching seafile
seafile       | 
seafile       | Starting seafile server, please wait ...
seafile       | ** Message: 17:47:00.629: seafile-controller.c(621): No seafevents.
seafile       | 
seafile       | Seafile server started
seafile       | 
seafile       | Done.
seafile       | 
seafile       | LANG is not set in ENV, set to en_US.UTF-8
seafile       | LC_ALL is not set in ENV, set to en_US.UTF-8
seafile       | ./seafile-server-latest/seahub.sh: line 220: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
seafile       | Starting seahub at port 8000 ...
seafile       | 
seafile       | ----------------------------------------
seafile       | It's the first time you start the seafile server. Now let's create the admin account
seafile       | ----------------------------------------
seafile       | 
seafile       | What is the email for the admin account?
seafile       | [ admin email ] 
seafile       | Error happened during creating seafile admin.
seafile       | 
seafile       | 
seafile       | Seahub is started
seafile       | 
seafile       | Done.
seafile       | 
seafile       | 2022-01-23T17:47:12+01:00 [Launch] Done
seafile       | 2022-01-23T17:47:12+01:00 [Entrypoint] Waiting for termination
```

**Comment by olof-nord** [comment]
To follow up on the logs, here is a section of the logs from the DB during the failed/semi failed initialisation.

```
seafile-db    | 2022-01-23 17:47:00 3 [Warning] Access denied for user 'seafile'@'172.31.0.4' (using password: NO)
seafile-db    | 2022-01-23 17:47:00 4 [Warning] Access denied for user 'seafile'@'%' to database 'seafile_db'
seafile-db    | 2022-01-23 17:47:00 5 [Warning] Access denied for user 'seafile'@'%' to database 'seafile_db'
seafile-db    | 2022-01-23 17:47:04 6 [Warning] Access denied for user 'seafile'@'%' to database 'ccnet_db'
seafile-db    | 2022-01-23 17:47:19 7 [Warning] Access denied for user 'seafile'@'%' to database 'seahub_db'
```
It seems in my case to be some permission issue?

**Comment by ChatDeBlofeld** [comment]
>Should the "seafile" user be created by Seafile, or exist before the init script?
>I understand that the Seafile databases (ccnet, seafile and seahub) are per default created during setup.

User's created at setup in the same script as the db ([here](https://github.com/haiwen/seahub/blob/master/scripts/setup-seafile-mysql.py#L649)).

Then what happens is the following:
- At every Seahub launch, [`check_init_admin.py` script](https://github.com/haiwen/seahub/blob/d6e4a60c62fb8954a8a10dc7193a444789b399e9/scripts/check_init_admin.py) checks the db to know if any admin is present (which returns `false` if there's a db issue).
- Script then looks for an `admin.txt` file containing the credentials. This file is generated by the [init script](https://github.com/ChatDeBlofeld/seafile-arm-docker-base/blob/master/scripts/init.sh#L161) at initialization and automatically removed afterwards.
- Since the file does not exist anymore after initial setup, it launches an interactive prompt which _should_ fail since it isn't an interactive environment. 
- Admin creation failure is logged and Seahub is launched anyway. 

What comes from this process is admin creation _can_ fail without being an issue: for example if db wasn't started at admin check but becomes reachable afterwards.

In summary: this error can be minor (but still unexpected, as written before I don't understand where it comes from) or a symptom of a bigger issue. If everything's working, there's nothing to worry too much about. Else, well...

>It seems in my case to be some permission issue?

I think so. Can you open another issue for this specific case? My advise would be to take advantage of the docker tools and wipe everything to start on clean basis. This obviously depends on your installation state (are there data stored, etc.)

**Comment by olof-nord** [comment]
Thanks for the detailed explanation! It turns out my issue was that I provided a MariaDB/MySQL user and password during docker DB startup, which should not be set there but instead are managed by the Seafile init, which adds a database user and password itself.

For MariaDB/MySQL only one environment variable is needed, and that's the root password (MARIADB_ROOT_PASSWORD/MYSQL_ROOT_PASSWD). The other error messages were not relevant.
